pkg/util/provider/machinecontroller/migrate_machine_class.go

```diff
@@ -0,0 +1,409 @@
+	// GCPMachineClassKind is used to identify the machineClassKind as GCP
+	GCPMachineClassKind = "GCPMachineClass"
+	// OpenStackMachineClassKind is used to identify the machineClassKind as OpenStack
+	OpenStackMachineClassKind = "OpenStackMachineClass"
+	// PacketMachineClassKind is used to identify the machineClassKind as Packet
+	PacketMachineClassKind = "PacketMachineClass"
+	// MachineController is the string constant to identify the controller responsible for the migration
+	MachineController = "machine-controller"
+)
+
+// getProviderSpecificMachineClass returns the providerSpecificMachineClass object for the provided machineClass
+func (c *controller) getProviderSpecificMachineClass(classSpec *v1alpha1.ClassSpec) (interface{}, error) {
+	var (
+		machineClass interface{}
+		err          error
+	)
+
+	switch classSpec.Kind {
+	case AlicloudMachineClassKind:
+		machineClass, err = c.controlMachineClient.AlicloudMachineClasses(c.namespace).Get(classSpec.Name, metav1.GetOptions{
+			TypeMeta:        metav1.TypeMeta{},
+			ResourceVersion: "",
+		})
+	case AWSMachineClassKind:
+		machineClass, err = c.controlMachineClient.AlicloudMachineClasses(c.namespace).Get(classSpec.Name, metav1.GetOptions{
+			TypeMeta:        metav1.TypeMeta{},
+			ResourceVersion: "",
+		})
+	case AzureMachineClassKind:
+		machineClass, err = c.controlMachineClient.AzureMachineClasses(c.namespace).Get(classSpec.Name, metav1.GetOptions{
+			TypeMeta:        metav1.TypeMeta{},
+			ResourceVersion: "",
+		})
+	case GCPMachineClassKind:
+		machineClass, err = c.controlMachineClient.GCPMachineClasses(c.namespace).Get(classSpec.Name, metav1.GetOptions{
+			TypeMeta:        metav1.TypeMeta{},
+			ResourceVersion: "",
+		})
+	case OpenStackMachineClassKind:
+		machineClass, err = c.controlMachineClient.OpenStackMachineClasses(c.namespace).Get(classSpec.Name, metav1.GetOptions{
+			TypeMeta:        metav1.TypeMeta{},
+			ResourceVersion: "",
+		})
+	case PacketMachineClassKind:
+		machineClass, err = c.controlMachineClient.PacketMachineClasses(c.namespace).Get(classSpec.Name, metav1.GetOptions{
+			TypeMeta:        metav1.TypeMeta{},
+			ResourceVersion: "",
+		})
+	default:
+		return nil, fmt.Errorf("Machine class kind not found")
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return machineClass, nil
+}
+
+// createMachineClass creates the generic machineClass corresponding to the providerMachineClass
+func (c *controller) createMachineClass(providerSpecificMachineClass interface{}, classSpec *v1alpha1.ClassSpec) (machineutils.Retry, error) {
+
+	machineClass, err := c.machineClassLister.MachineClasses(c.namespace).Get(classSpec.Name)
+	if err != nil && apierrors.IsNotFound(err) {
+		// MachineClass doesn't exist, initialize a new one
+		machineClass = &v1alpha1.MachineClass{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: classSpec.Name,
+			},
+		}
+		machineClass, err = c.controlMachineClient.MachineClasses(c.namespace).Create(machineClass)
```

Comment:
>  have only one small comment about if it is possible to create the migrated MachineClass in one go rather than in two steps (Create followed up Update).

Hi Amshu, 

I was facing some complications while doing the same. I am going to leave it as-is for, as it shouldn't cause too many calls as it is called only during migrations. I hope that's okay.